book/src/solana-scripts.md

```diff
@@ -0,0 +1,253 @@
+order.
+
+This design proposes another approach.  Users know ahead of time which
+instructions the script will generate, and if signatures are required for the
+instructions.
+
+The transaction invoked by the client must declare all the accounts that the
+script will need up front and provide all the necessary client signatures, as
+well as encode the instruction vector that the script will generate.
+
+The latter provides clear authorization for the script to take actions on behalf
+of the client, since each instruction specifies the clients keys and explicit
+signatures for each explicit instruction.  Users do not need to guess which
+instructions the script will execute, and authorize each one explicitly.  If the
+on-chain state has changed such that the script will take a different branch and
+execute a different instruction, then the script fails.  This allows the
+programs to implement out of order composition and for users to take advantage
+of those performance benefits.
+
+Once the instruction is invoked, the script is resumed from the last point of
+execution.
+
+## Methods
+
+A TTP needs the ability to create pubkeys and generate signatures for those
+keys.  At script creation time the loader program may authorize specific
+pubkeys that the script and only the script can sign with.  To ensure that these
+keys cannot be signed by the client, the addresses are derived from a sha256 of
+the script pubkey and the key sequence number.
+
+* `pub fn keypair_pubkey(key_index: u64) -> Pubkey`
+
+Retrivies the script keypair at index `key_index`.  The pubkey of the keypair is
+`sha256.hash(program_id).hash(key_index)`, and therefore has no real private
+key.  This keypair can only be used to sign messages with the `sign_instruction`
+function by the script.
+
+* `pub fn sign_instruction(ix: &mut Instruction, key_index: u64) -> ()`
+
+Signs the message for the script pubkey that is generated with `key_index`.
+Clients can generate these keys locally and encode them into the instruction
+vector.  Only scripts can sign with these keys.  While a client is expected to
+encode the key into the instruction vector, the execution of the script will
+will call `sign_instruction` and set the `KeyedAccounts::is_signer` flag.
+
+* `fn process_instruction(ix: Instruction) -> Result<(), InstructionError>`
+
+This method is available to scripts to execute an instruction in the runtime.
+
+## Scirpt Initialization
+
+* `LoaderInstruction::FinalizeScript`
+
+`LoaderInstruction::FinalizeScript` designates that the loaded executable
+bytecode is a script, and creates a new instance of the script. The difference
+between scripts and programs is that script execution yields to external program
+instructions, and scripts have the capability to sign.  `FinalizeScript` may be
+called more than once on the same loaded bytecode to create unique instances of
+scripts each with their own signing keys.
+
+## Script Instruction Vector
```

Comment:
Does the runtime enforce that the script is issuing the instructions in the vector?

Are the instructions cached and submitted by the runtime all at once at script completion?

Reply:
@jackcmay the runtime checks that the instructions issued by the script match what is encoded in the transaction.  if the addresses or user data doesn't match, the transaction fails.

Reply:
So the output of the script is a transaction that matches the `script instruction vector'?

Reply:
@jackcmay user knows a head of time what instructions the script will issue, and encodes them into the TX.  The script then executes, and the execution path must match what the user specified.

Reply:
The goal is to allow safe unordered execution between users of the same program.  Otherwise, think of a N player game without turns, like the exchange :).

Reply:
The program you refer to in your last comment is actually the script itself and not the programs that scripts calls?

Reply:
@jackcmay yea, fixed.  if we have a single program, thats easy, but as soon as you try to compose two different programs that have multiple entry points I think the number of state permutations blows up to an unmanageable amount.